README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# DNA Methylation Arrays Processing And Quality-Control

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![GitHub
tag](https://img.shields.io/github/tag/umr1283/dmapaq.svg?label=latest%20tag&include_prereleases)](https://github.com/umr1283/dmapaq)
[![R build
status](https://github.com/umr1283/dmapaq/workflows/R-CMD-check/badge.svg)](https://github.com/umr1283/dmapaq/actions)
<!-- badges: end -->

## Installation

``` r
# Install dmapaq from CRAN:
install.packages("dmapaq")

# Or the the development version from GitHub:
# install.packages("remotes")
remotes::install_github("umr1283/dmapaq")
```

## Overview

-   `read_idats()` allows to efficiently import idats files mostly using
    [minfi](https://bioconductor.org/packages/minfi/) functions.
-   `qc_idats()` allows to compute quality-control of methylation array
    from Illumina using a [rmarkdown
    template](inst/rmarkdown/templates/qc_idats/skeleton/skeleton.Rmd).

------------------------------------------------------------------------

## Getting help

If you encounter a clear bug, please file a minimal reproducible example
on [github](https://github.com/umr1283/dgapaq/issues).  
For questions and other discussion, please contact the package
maintainer.

## Code of Conduct

Please note that the `dgapaq` project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# dmapaq (development version)

* In `DESCRIPTION`,
    + Add Illumina annotation in GRCh38.
* In [rmarkdown templates](inst/rmarkdown/templates),
    + Tweak code syntax.
    + Add Illumina annotation in GRCh38.
    + Fix cell composition estimate for CordBloodCombined.
* In `read_idats.R`,
    + Fix `.doSnowGlobals` errors within `targets`.
    + Add Illumina annotation in GRCh38 as default.

# dmapaq 0.3.5

* In `DESCRIPTION`,
    + Remove `doParallel`.
* In [rmarkdown templates](inst/rmarkdown/templates),
    + Make sure `Sample_ID` is a unique string and in the first column of the sample sheet.
    + Remove `doParallel`.
    + Order PCA association tile plot according to p-values.
* In `read_idats.R`,
    + Rewrite the build process of the `RGChannelSetExtended` object.

# dmapaq 0.3.4

* In [rmarkdown templates](inst/rmarkdown/templates),
    + Explicit use of `Sample_ID` as the unique IDs for samples.
* In `read_idats.R`,
    + Explicit use of `Sample_ID` as the unique IDs for samples.

# dmapaq 0.3.3

* In [rmarkdown templates](inst/rmarkdown/templates),
    + Update obsolete captions.
    + Perform independent association testing between technical variables and principal components.
    + Fix `ggplot2` issue with legend components defined in `theme_set()`.

# dmapaq 0.3.2

* In [rmarkdown templates](inst/rmarkdown/templates),
    + Remove unnecessary lower condition for outliers.
    + Replace underscore with dash.
    + Fix PCA section.

# dmapaq 0.3.1

* In [rmarkdown templates](inst/rmarkdown/templates),
    + Fix reference for ComBat.
    + Fix missing params in YAML.
* In `read_idats.R`,
    + Fix dataset for population specific CpG/SNPs.
    + Fix missing params in YAML.

# dmapaq 0.3.0

* In [rmarkdown templates](inst/rmarkdown/templates),
    + Reorder code in setup chunk.
    + Rewrite code using `data.table`.
* In `ggheatmap.R`,
    + Remove function `ggheatmap()`.
* In `read_idats.R`,
    + Optimise code.
    + Use `parallel`.
    + Use `data.table`.
    + Add `echo` parameter.
    + Return log messages.

# dmapaq 0.2.0

## Minor improvements and fixes

* In `DESCRIPTION`, 
    + Update all packages listed in `Imports`.
    + Now imports `ggplot2 v3.3.0`.
    + Use `gt` for tables.
* In `README.Rmd`, remove parallel function for ComBat batch correction.
* In [rmarkdown template](inst/rmarkdown/templates/qc_idats/skeleton/skeleton.Rmd),
    + Fix regular expression for reference panel.
    + Fix library availability check for cordbloodcombined reference panel.
    + Fix misplaced parenthesis around calls to `system.file()`.
    + Fix missing probe id in beta matrix output.
    + Fix missing prefix.
    + Fix cell composition section when no tissue is provided.
    + Fix gender discrepancy table when observed gender is missing.
    + Fix call rates table not displaying all thresholds.
    + Remove parallel computation for ComBat batch correction.
    + Replace `requireNamespace()` with `system.file()` in order to check packages availability.
    + No longer show empty gender check table.
* In `R/qc_idats.R`.
    + Import `gt` for tables.
    + Remove import for `kableExtra`.
* Remove `R/ComBat.mc`.

# dmapaq 0.1.0

## New features

* `read_idats()` allows to efficiently import idats files mostly 
    using [minfi](https://bioconductor.org/packages/minfi/) functions.
* `qc_idats()` allows to compute quality-control of methylation array from Illumina 
    using a [rmarkdown template](inst/rmarkdown/templates/qc_idats/skeleton/skeleton.Rmd).
* `ggheatmap()` allows to compute heatmap with dendrogram on x-axis and y-axis 
    using [ggplot2](https://ggplot2.tidyverse.org/).
* `ComBat.mc()` is a multi-processor wrapper for ComBat method 
    (initially in [ENmix](https://bioconductor.org/packages/ENmix/)).


tests/testthat.R
library(testthat)
library(dmapaq)

# test_check("dmapaq")


R/ComBat.mc.R
#' A multi-processor wrapper for ComBat method.
#'
#' A multi-processor wrapper for ComBat method.
#' ComBat is a method to adjust batch effect where the batch covariate is known.
#'
#' @param dat A data matrix with column for samples and row for probe.
#' @param batch Batch covariate (multiple batches allowed)
#' @param nCores Number of cores will be used for computation
#' @param ... See `ComBat` in `sva` package for extra options
#'
#' @return A data matrix with the same dimension as input data, adjusted for batch effects.
#'     Warning: Values for multimodal distributed CpGs could be over-adjusted.
#'
#' @references Johnson, WE, Rabinovic, A, and Li, C (2007).
#'     *Adjusting batch effects in microarray expression data using Empirical Bayes methods. Biostatistics 8(1):118-127.*
#'
#' @seealso See `ComBat` in `sva` package for extra options
#'
#' @importFrom foreach `%dopar%`
#'
#' @export
ComBat.mc <- function(dat, batch, nCores = 1, ...) {
  #    if(!require("sva")){stop("Can not load sva package")}
  if (nCores > parallel::detectCores()) {
    nCores <- parallel::detectCores()
    cat(
      "Only ", parallel::detectCores(), " Cores avialable, nCores was reset to ",
      parallel::detectCores(), "\n"
    )
  }
  cat("Analysis is running, please wait...!", "\n")

  rname <- rownames(dat)
  nparts1 <- min(ceiling(ncol(dat) / 500), floor(nrow(dat) / 40000))
  parts1 <- rep(1:nparts1, ceiling(nrow(dat) / nparts1))[1:nrow(dat)]
  nCores <- min(10, nCores, ceiling(nrow(dat) / (nparts1 * 20000)))
  dat.o <- NULL
  c1 <- parallel::makeCluster(nCores)
  doParallel::registerDoParallel(c1)
  for (i in 1:nparts1) {
    dat1 <- dat[which(parts1 == i), ]
    parts <- rep(1:nCores, ceiling(nrow(dat1) / nCores))[1:nrow(dat1)]
    parts <- sample(parts)
    dat.o1 <- foreach::foreach(s = 1:nCores, .combine = rbind, .export = c("sva::ComBat")) %dopar% {
      s <- s
      idx <- which(parts == s)
      sva::ComBat(dat = dat1[idx, ], batch = batch, ...)
    }
    dat.o <- rbind(dat.o, dat.o1)
  }
  parallel::stopCluster(c1)
  dat.o[rname, ]
}


R/read_idats.R
#' Efficiently import idats files mostly using minfi functions.
#'
#' @param directory A `character`. Location of IDAT files, default is the current working directory.
#' @param csv_file A `character`. Path to the sample sheet (csv files) or
#'     name of the sample sheet in `directory`.
#' @param meth_value_type A `character`. Indicates whether you prefer m-values (`"M"`)
#'     or beta-values (`"B"`). Default is `"B"`.
#' @param array_name A `character`. Choose microarray type, eiyther `"450K"` or `"EPIC"`.
#'     Default is `"EPIC"`.
#' @param annotation_version A `character`. Version of the annotation package that should be used.
#'     Default is `"ilm10b4.hg19"` for the `"EPIC"` array
#' @param n_cores An `integer`. The number of cores to use,
#'     i.e., at most how many child processes will be run simultaneously.
#' @param rgSet A `RGChannelSet` object.
#' @param echo A `logical`. Should messages be displayed?
#'
#' @inheritParams qc_idats
#'
#' @import data.table
#' @import ChAMPdata
#'
#' @return A `list`.
#' @export
read_idats <- function(
  directory = getwd(),
  csv_file = "csv$",
  meth_value_type = "B",
  filter_beads = TRUE,
  bead_cutoff = 0.05,
  filter_non_cpg = TRUE,
  filter_snps = TRUE,
  population = NULL,
  filter_multihit = TRUE,
  filter_xy = TRUE,
  detection_pvalues = 0.01,
  filter_callrate = TRUE,
  callrate_samples = 0.99,
  callrate_probes = 1,
  norm_background = "oob",
  norm_dye = "RELIC",
  norm_quantile = "quantile1",
  array_name = c("EPIC", "450k"),
  annotation_version = c("ilm10b5.hg38", "ilm10b4.hg19", "ilmn12.hg19"),
  n_cores = 1,
  rgSet = NULL,
  echo = FALSE
) {

  old_r_libs <- Sys.getenv("R_LIBS")
  r_libs <- unique(c(
    .libPaths()[1],
    unlist(strsplit(Sys.getenv("R_LIBS"), .Platform$path.sep))
  ))
  Sys.setenv(R_LIBS = paste(r_libs, collapse = .Platform$path.sep))
  on.exit(Sys.setenv(R_LIBS = old_r_libs))

  if (echo) {
    message(
      "===============================\n",
      "[dmapaq] Reading IDAT files ...\n",
      "===============================",
      appendLF = TRUE
    )
  }

  array_name <- array_name[1]
  annotation_version <- annotation_version[1]

  stopifnot(suppressPackageStartupMessages(
    nchar(system.file(package = "minfi")) > 0 &
      switch(
        EXPR = array_name,
        "450k" = nchar(system.file(package = "IlluminaHumanMethylation450kmanifest")) > 0,
        "EPIC" = nchar(system.file(package = "IlluminaHumanMethylationEPICmanifest")) > 0
      )
  ))

  if (is.null(rgSet) | !inherits(rgSet, "RGChannelSet")) {
    sample_sheet <- read_sample_sheet(directory = directory, csv_file = csv_file, echo = echo)
    rgSet <- read_metharray_exp(sample_sheet = sample_sheet, n_cores = n_cores)
  }
  rgSet@annotation <- switch(
    EXPR = array_name,
    "450k" = c(array = "IlluminaHumanMethylation450k", annotation = annotation_version),
    "EPIC" = c(array = "IlluminaHumanMethylationEPIC", annotation = annotation_version)
  )

  data_detP <- minfi::detectionP(rgSet)
  data_detP[is.na(data_detP)] <- 1

  if (filter_callrate) {
    good_detection <- data_detP < detection_pvalues

    call_rate_samples <- colSums(good_detection) / nrow(good_detection)
    bad_samples <- names(which(call_rate_samples < callrate_samples))

    good_detection <- good_detection[, setdiff(colnames(good_detection), bad_samples)]

    call_rate_cpg <- rowSums(good_detection) / ncol(good_detection)
    bad_cpgs <- names(which(call_rate_cpg < callrate_probes))
  } else {
    bad_samples <- NULL
    bad_cpgs <- NULL
  }

  mset_raw <- mset <- minfi::preprocessRaw(rgSet)

  if (filter_beads) {
    bc <- get_beadcount(rgSet)
    bc2 <- bc[rowSums(is.na(bc)) < bead_cutoff * (ncol(bc)), ]
    mset_f2 <- mset[minfi::featureNames(mset) %in% rownames(bc2), ]
    n_beads_discarded <- format(dim(mset)[1] - dim(mset_f2)[1], big.mark = ",", digits = 0)
    mset <- mset_f2
  }

  if (filter_non_cpg) {
    mset_f2 <- minfi::dropMethylationLoci(mset, dropCH = TRUE)
    n_non_cpg_discarded <- format(dim(mset)[1] - dim(mset_f2)[1], big.mark = ",", digits = 0)
    mset <- mset_f2
  }

  if (filter_snps) {
    ref_population <- c(
      "AFR", "EAS", "EUR",
      "SAS", "AMR", "GWD", "YRI", "TSI", "IBS",
      "CHS", "PUR", "JPT", "GIH", "CHB", "STU",
      "ITU", "LWK", "KHV", "FIN", "ESN", "CEU",
      "PJL", "ACB", "CLM", "CDX", "GBR", "BEB",
      "PEL", "MSL", "MXL", "ASW"
    )

    if (is.null(population) || !(population %in% ref_population)) {
      manifest_hg19 <- switch(
        EXPR = array_name,
        "450k" = get(utils::data("hm450.manifest.hg19", package = "ChAMPdata")),
        "EPIC" = get(utils::data("EPIC.manifest.hg19", package = "ChAMPdata"))
      )
      which_population <- which(manifest_hg19$MASK_general)
    } else {
      manifest_hg19 <- switch(
        EXPR = array_name,
        "450k" = get(utils::data("hm450.manifest.pop.hg19", package = "ChAMPdata")),
        "EPIC" = get(utils::data("EPIC.manifest.pop.hg19", package = "ChAMPdata"))
      )
      which_population <- which(manifest_hg19[, paste("MASK_general", population, sep = "_")])
    }
    maskname <- rownames(manifest_hg19)[which_population]
    mset_f2 <- mset[!minfi::featureNames(mset) %in% maskname, ]
    n_snps_discarded <- format(dim(mset)[1] - dim(mset_f2)[1], big.mark = ",", digits = 0)
    mset <- mset_f2
  }

  if (filter_multihit) {
    multi_hit <- get(utils::data("multi.hit", package = "ChAMPdata"))
    mset_f2 <- mset[!minfi::featureNames(mset) %in% multi_hit$TargetID, ]
    n_multihit_discarded <- format(dim(mset)[1] - dim(mset_f2)[1], big.mark = ",", digits = 0)
    mset <- mset_f2
  }

  if (filter_xy) {
    switch(
      EXPR = array_name,
      "450k" = utils::data("probe.features", package = "ChAMPdata"),
      "EPIC" = utils::data("probe.features.epic", package = "ChAMPdata")
    )
    probe_features <- get("probe.features")
    autosomes <- probe_features[!probe_features$CHR %in% c("X", "Y"), ]
    mset_f2 <- mset[minfi::featureNames(mset) %in% rownames(autosomes), ]
    n_xy_discarded <- format(dim(mset)[1] - dim(mset_f2)[1], big.mark = ",", digits = 0)
    mset <- mset_f2
  }

  mset <- ENmix::preprocessENmix(
    rgSet = rgSet,
    bgParaEst = norm_background,
    dyeCorr = norm_dye,
    QCinfo = NULL,
    exQCsample = FALSE,
    exQCcpg = FALSE,
    exSample = bad_samples,
    exCpG = unique(c(bad_cpgs, setdiff(minfi::featureNames(mset_raw), minfi::featureNames(mset)))),
    nCores = n_cores
  )
  mset <- ENmix::norm.quantile(mdat = mset, method = norm_quantile)

  methylation_matrix <-switch(meth_value_type,
    "B" = minfi::getBeta(mset, "Illumina"),
    "M" = minfi::getM(mset),
    stop('Methylation value type not defined. Only "B" or "M" are available.')
  )

  if (min(methylation_matrix, na.rm = TRUE) <= 0) {
    methylation_matrix[methylation_matrix <= 0] <- min(methylation_matrix[methylation_matrix > 0], na.rm = TRUE)
  }
  if (max(methylation_matrix, na.rm = TRUE) >= 1) {
    methylation_matrix[methylation_matrix >= 1] <- max(methylation_matrix[methylation_matrix < 1], na.rm = TRUE)
  }

  colnames(methylation_matrix) <- minfi::pData(mset)[["Sample_ID"]]
  mset@metadata[[meth_value_type]] <- methylation_matrix
  tmp_phenotypes <- data.table::as.data.table(minfi::pData(rgSet))
  tmp_phenotypes[, "Sample_ID" := lapply(.SD, as.character), .SDcols = "Sample_ID"]
  tmp_means <- colMeans(data_detP)[tmp_phenotypes[["Sample_ID"]]]
  tmp_phenotypes[, "mean_detection_pvalue" := tmp_means]
  tmp_callrate <- (colSums(data_detP < detection_pvalues) / nrow(data_detP))[tmp_phenotypes[["Sample_ID"]]]
  tmp_phenotypes[, "call_rate" := tmp_callrate]
  mset@metadata[["phenotypes"]] <- tmp_phenotypes

  log_msg <- character(0)
  if (filter_callrate) {
    log_msg <- c(log_msg, paste0(
      "Filtering samples with call rate below ",
      paste(format(callrate_samples * 100, digits = 1, nsmall = 1), "%"), ":\n",
      "  - ", format(length(bad_samples), big.mark = ",", digits = 0), " samples were discarded."
    ))
    log_msg <- c(log_msg, paste0(
      "Filtering probes with call rate below ",
      paste(format(callrate_probes * 100, digits = 1, nsmall = 1), "%"), ":\n",
      "  - ", format(length(bad_cpgs), big.mark = ",", digits = 0), " probes were discarded."
    ))
  }
  if (filter_beads) {
    log_msg <- c(log_msg, paste0(
      "Filtering probes with a beadcount lower than three in at least ",
      paste(format(bead_cutoff * 100, digits = 1, nsmall = 1), "%"), " of samples:\n",
      "  - ", n_beads_discarded, " probes were discarded."
    ))
  }
  if (filter_non_cpg) {
    log_msg <- c(log_msg, paste0(
      "Filtering non-cg probes:\n",
      "  - ", n_non_cpg_discarded, " probes were discarded."
    ))
  }
  if (filter_snps) {
    log_msg <- c(log_msg, paste0(
      "Filtering probes with SNPs (Zhou et al., 2016; doi:10.1093/nar/gkw967):\n",
      "  - ", n_snps_discarded, " probes were discarded."
    ))
  }
  if (filter_multihit) {
    log_msg <- c(log_msg, paste0(
      "Filtering probes that align to multiple locations ",
      "(Nordlund et al., 2013; doi:10.1186/gb-2013-14-9-r105):\n",
      "  - ", n_multihit_discarded, " probes were discarded."
    ))
  }
  if (filter_xy) {
    log_msg <- c(log_msg, paste0(
      "Filtering probes on the X or Y chromosome:\n",
      "  - ", n_xy_discarded, " probes were discarded."
    ))
  }
  log_msg <- c(log_msg,
    "Zeros have been replaced with smallest value over zero.",
    "Ones have been replaced with largest value below one.",
    paste0(
      "Data contains:\n",
      "  - ", format(dim(methylation_matrix)[1], big.mark = ",", digits = 0), " probes.\n",
      "  - ", format(dim(methylation_matrix)[2], big.mark = ",", digits = 0), " samples.\n",
      "  - ", format(sum(is.na(methylation_matrix)), big.mark = ",", digits = 0), " missing values."
    )
  )
  if (echo) {
    message(paste(log_msg, collapse = "\n"), appendLF = TRUE)
  }

  list(mset = mset, rgset = rgSet, log = log_msg)
}


#' read_sample_sheet
#'
#' @inheritParams read_idats
#' @param ignore.case A `logical`. A logical value.
#'     If `TRUE`, the directory path is prepended to the file names to give a relative file path.
#'     If `FALSE`, the file names (rather than paths) are returned.
#' @param recursive A `logical`. Should the listing recurse into directories?
#' @param full.names A `logical`. Should pattern-matching be case-insensitive?
#'
#' @keywords internal
read_sample_sheet <- function(
  directory = ".",
  csv_file = "csv$",
  ignore.case = TRUE,
  recursive = TRUE,
  full.names = TRUE,
  echo = FALSE
) {
  if (file.exists(suppressWarnings(normalizePath(csv_file)))) {
    list_files <- normalizePath(csv_file)
  } else {
    list_files <- list.files(
      path = directory,
      pattern = csv_file,
      full.names = full.names,
      ignore.case = ignore.case,
      recursive = recursive
    )
    if (length(list_files)>1) {
      warning("[dmapaq] More than one CSV file have been found!")
      list_files <- list.files[1]
      if (echo) message("[dmapaq] File '", list.files, "' will be used.", appendLF = TRUE)
    }
  }

  data_header <- grep("^\\[DATA\\]", readLines(list_files), ignore.case = TRUE)
  if (length(data_header) == 0) data_header <- 0
  col_names <- colnames(data.table::fread(file = list_files, skip = data_header, nrows = 1))
  default_cols <- c("Sample_ID", "Sentrix_ID", "Sentrix_Position")

  cols_missing <- default_cols[!default_cols %in% col_names]
  if (length(cols_missing) != 0) {
    stop(
      "[dmapaq] Sample Sheet must contains the following missing columns:\n",
      "  - ", paste(cols_missing, collapse = "\n  - ")
    )
  }

  sample_sheet <- data.table::fread(file = list_files, skip = data_header)
  data.table::setnames(
    x = sample_sheet,
    old = c("Sample_ID", "Slide", "Array", "Sample_Plate", "Sample_Well"),
    new = c("Sample_ID", "Sentrix_ID", "Sentrix_Position", "Sample_Plate", "Sample_Well"),
    skip_absent = TRUE
  )

  basenames <- sub("_Grn\\.idat.*", "", sapply(
    X = paste0(sample_sheet[["Sentrix_ID"]], "_", sample_sheet[["Sentrix_Position"]], "_Grn.idat"),
    FUN = grep,
    x = list.files(path = directory, recursive = recursive, full.names = TRUE),
    value = TRUE,
    USE.NAMES = FALSE
  ), ignore.case = TRUE)
  sample_sheet[, "Basename" := basenames]

  sample_sheet
}

#' read_metharray
#'
#' @inheritParams read_idats
#' @param files The `basenames` or `filenames` of the IDAT files.
#'     `basenames` are the filename without the ending `_Grn.idat` or `_Red.idat`.
#'     `filenames` are filenames including `_Grn.idat` or `_Red.idat`.
#'
#' @keywords internal
read_metharray <- function(files, n_cores) {
  Grn <- Red <- NULL # to avoid note from global variable check
  basenames <- unique(sub("_Grn\\.idat.*|_Red\\.idat.*", "", files))
  for (ichannel in c("Grn", "Red")) {
    i_files <- paste0(basenames, "_", ichannel, ".idat")
    names(i_files) <- basename(basenames)
    i_files_exists <- file.exists(i_files)
    if (!all(i_files_exists)) {
      i_filesgz_exists <- file.exists(paste0(i_files, ".gz"))
      if (!all(i_filesgz_exists)) {
        stop(
           "The following specified files do not exist:\n",
          "  - ", paste(i_files[!i_files_exists], collapse = "\n  - ")
        )
      }
      i_files <- paste0(i_files, ".gz")
    }
    suppressWarnings({
      i_idats <- parallel::mclapply(
        X = i_files, mc.preschedule = FALSE, mc.cores = n_cores,
        FUN = illuminaio::readIDAT
      )
    })
    assign(x = ichannel, value = i_idats)
  }

  G_idats <- lapply(X = Grn, FUN = `[[`, "Quants")
  R_idats <- lapply(X = Red, FUN = `[[`, "Quants")

  common_rows <- as.character(Reduce("intersect", lapply(X = G_idats, FUN = rownames)))

  out <- minfi::RGChannelSetExtended(
    Red = do.call("cbind", lapply(X = R_idats, y = common_rows, FUN = function(x, y) x[y, "Mean"])),
    Green = do.call("cbind", lapply(X = G_idats, y = common_rows, FUN = function(x, y) x[y, "Mean"])),
    RedSD = do.call("cbind", lapply(X = R_idats, y = common_rows, FUN = function(x, y) x[y, "SD"])),
    GreenSD = do.call("cbind", lapply(X = G_idats, y = common_rows, FUN = function(x, y) x[y, "SD"])),
    NBeads = do.call("cbind", lapply(X = G_idats, y = common_rows, FUN = function(x, y) x[y, "NBeads"]))
  )

  rownames(out) <- common_rows

  out
}

#' read_metharray_exp
#'
#' @inheritParams read_idats
#' @inheritParams read_metharray
#' @inheritParams read_sample_sheet
#'
#' @keywords internal
read_metharray_exp <- function(
  directory = NULL,
  sample_sheet = NULL,
  ignore.case = TRUE,
  recursive = TRUE,
  full.names = TRUE,
  n_cores = 1
) {
  if (is.null(sample_sheet)) {
    if (is.null(directory)) directory <- "."
    green_files <- list.files(
      path = directory,
      pattern = "_Grn.idat.*",
      recursive = recursive,
      ignore.case = ignore.case,
      full.names = full.names
    )
    red_files <- list.files(
      path = directory,
      pattern = "_Red.idat.*",
      recursive = recursive,
      ignore.case = ignore.case,
      full.names = full.names
    )

    if (length(green_files) == 0 || length(red_files) == 0) {
      stop("IDAT files must be provided.")
    }

    common_files <- intersect(sub("_Grn.idat.*", "", green_files), sub("_Red.idat.*", "", red_files))

    if (length(common_files) == 0) {
      stop('"Grn" and "Red" idats files must be provided.')
    }

    common_files_green <- paste0(common_files, "_Grn.idat")
    if (!setequal(common_files_green, green_files)) {
      warning(
        "The following files only exists for the green channel:\n",
        "  - ", paste(setdiff(green_files, common_files_green), collapse = "\n  - ")
      )
    }

    common_files_red <- paste0(common_files, "_Red.idat")
    if (!setequal(common_files_red, red_files)) {
       warning(
        "The following files only exists for the red channel:\n",
        "  - ", paste(setdiff(red_files, common_files_red), collapse = "\n  - ")
      )
    }

    rgSet <- read_metharray(common_files, n_cores)
  } else {
    if (all(!grepl("Basename", names(sample_sheet)))) {
      stop('"Basename" must be provided as a column of "sample_sheet".')
    }

    if (is.null(directory)) {
      files <- sample_sheet[["Basename"]]
    } else {
      files <- file.path(directory, basename(sample_sheet[["Basename"]]))
    }

    rgSet <- read_metharray(files, n_cores)

    pD <- methods::as(sample_sheet, "DataFrame")
    pD[["filenames"]] <- files
    rownames(pD) <- colnames(rgSet)
    rgSet@colData <- pD
  }

  minfi::sampleNames(rgSet) <- rgSet[["Sample_ID"]]

  rgSet
}

#' get_beadcount
#'
#' @param x `rgSet`
#'
#' @keywords internal
get_beadcount <- function(x) {
  nb <- minfi::getNBeads(x)
  typeI <- minfi::getProbeInfo(x, type = "I")
  typeII <- minfi::getProbeInfo(x, type = "II")
  locus_names <- minfi::getManifestInfo(x, "locusNames")

  bc_temp <- matrix(
    data = NA_real_,
    ncol = ncol(x),
    nrow = length(locus_names),
    dimnames = list(locus_names, minfi::sampleNames(x))
  )
  bc_temp[typeII$Name, ] <- nb[typeII$AddressA, ]
  bc_temp[typeI$Name, ] <- nb[typeI$AddressB, ]
  bc_temp[typeI$Name, ] <- nb[typeI$AddressA, ]
  bc_temp[which(nb[typeI$AddressA, ] < 3 | nb[typeI$AddressB, ] < 3)] <- NA

  data.frame(bc_temp)
}


R/qc_idats.R
#' Compute quality-control of methylation array from Illumina using a rmarkdown template.
#'
#' @param csv_file A `character`. The path to a CSV file, *i.e.*, a sample sheet describing the data.
#' @param data_directory A `character`. The path to the data directory.
#' @param array A `character`. The array name, *i.e.*, `"EPIC"` or `"450k"`.
#' @param annotation A `character`. The name and version of the annotation package to be used.
#' @param cohort_name A `character`. The name of the studied cohort / population.
#' @param output_file A `character`. The name of the html file produced.
#' @param output_directory A `character`. The path to the output directory.
#' @param filter_snps A `logical`. Should the probes in which the probed CpG falls near a SNP
#'     (according to ([Zhou et al., 2016](https://www.doi.org/10.1093/nar/gkw967))) be removed?
#'     Default is `TRUE`.
#' @param filter_non_cpg A `logical`. Should the non-cg probes be removed?
#' @param filter_xy A `logical`. Should the probes from X and Y chromosomes be removed?
#'     Default is `TRUE`.
#' @param filter_multihit A `logical`. Should the probes which align to multiple locations
#'     (according to [Nordlund et al., 2013](https://www.doi.org/10.1186/gb-2013-14-9-r105)) be removed?
#'     Default is `TRUE`.
#' @param filter_beads A `logical`. Should the probes with a beadcount less than 3 be removed?
#'     Default is `TRUE`.
#' @param population A `character`. Name of the ethnicity population to be used.
#'     Default is `NULL` for none.
#' @param bead_cutoff A `numeric`. The threshold for beadcount.
#'     Default is `0.05`.
#' @param detection_pvalues A `numeric`. The threshold for the detection pvalues above which,
#'     values are considered as missing. Default is `0.01`.
#' @param filter_callrate A `logical`. Should the data be filtered based on call rate metric?
#'     Default is `TRUE`.
#' @param callrate_samples A `numeric`. The call rate threshold for samples, under which samples are excluded.
#'     Default is `0.99`.
#' @param callrate_probes A `numeric`. The call rate threshold for probes, under which probes are excluded.
#'     Default is `1`.
#' @param gender_threshold A `numeric`. The threshold value to discrimate gender based
#'     on sexual chromosomes methylation.
#'    Default is `-2`.
#' @param gender_colname A `character`. The name of the column containing the gender
#'     in the file provided in `csv_file`.
#'     Default is `NULL`.
#' @param norm_background A `character`. Optional method to estimate background normal distribution parameters.
#'     This must be one of `"oob"`, `"est"` or `"neg"`.
#'     Default is `"oob"`.
#' @param norm_dye A `character`. Dye bias correction, "mean": correction based on averaged red/green ratio;
#'     or `"RELIC"`: correction with RELIC method;
#'     or `"none"`: no dye bias correction.
#'     Default is `"RELIC"`.
#' @param norm_quantile A `character`. The quantile normalisation to be used.
#'     This should be one of `"quantile1"`, `"quantile2"`, or `"quantile3"`.
#'     Default is `"quantile1"`.
#' @param cell_tissue A `character`. The cell tissue to be used for cell composition estimation,
#'     using a reference panel (*i.e.*, `"blood"` or `"cordblood"`) or a mathematical deconvolution.
#'     Default is `NULL`.
#' @param pca A `logical`. Whether or not a PCA should be performed on the dataset.
#'     Default is `TRUE`.
#' @param pca_vars A `vector(character)`. Variables to be used with factorial planes.
#'     Default is `c("Sample_Plate", "Sentrix_ID")`.
#' @param pca_threshold A `numeric`. The number of times the interquartile range from the upper
#'     and lower quartile at which sa sample is defined as an outlier. Default is `2`.
#' @param max_labels A `numeric`. The maximum number of labels to show on plots. Default is `15`
#' @param title A `character`. The report's title. Default is `paste(array, "Array Quality-Control")`.
#' @param author_name A `character`. The author's name to be printed in the report.
#'     Default is `Unknown`.
#' @param author_affiliation A `character`. The affiliation to be printed in the report.
#'     Default is `NULL`.
#' @param author_email A `character`. The email to be printed in the report.
#'     Default is `NULL`.
#' @param encoding A `character`. The encoding to be used for the html report.
#'     Default is `"UTF-8"`.
#' @param ... Parameters to pass to `rmarkdown::render()`.
#'
#' @return NULL
#'
#' @export
qc_idats <- function(
  csv_file,
  data_directory,
  array = "EPIC",
  annotation = "ilm10b4.hg19",
  cohort_name = "COHORT",
  output_file = paste(cohort_name, array, "QC.html", sep = "_"),
  output_directory = ".",
  filter_snps = TRUE,
  filter_non_cpg = TRUE,
  filter_xy = TRUE,
  filter_multihit = TRUE,
  filter_beads = TRUE,
  population = NULL,
  bead_cutoff = 0.05,
  detection_pvalues = 0.01,
  filter_callrate = TRUE,
  callrate_samples = 0.99,
  callrate_probes = 1,
  gender_threshold = -2,
  gender_colname = NULL,
  norm_background = "oob",
  norm_dye = "RELIC",
  norm_quantile = "quantile1",
  cell_tissue = NULL,
  pca = TRUE,
  pca_vars = c("Sample_Plate", "Sentrix_ID"),
  pca_threshold = 2,
  max_labels = 15,
  title = paste(array, "Array Quality-Control"),
  author_name = "Unknown",
  author_affiliation = NULL,
  author_email = NULL,
  encoding = "UTF-8",
  ...
) {
  message_prefix <- "[dmapaq] "

  message(message_prefix, "Quality-Control started ...")

  invisible(file.copy(
    from = system.file("rmarkdown", "templates", "qc_idats", "skeleton.Rmd", package = "dmapaq"),
    to = file.path(tempdir(), "qc_idats.Rmd"),
    overwrite = TRUE
  ))

  if (!all(sapply(
    X = strsplit(gsub("\\([^()]+\\)|\n| ", "", utils::packageDescription("dmapaq", fields = "Suggests")), ",")[[1]],
    FUN = function(x) nchar(system.file(package = x)) > 0
  ))) {
    stop(message_prefix, "Packages listed in 'Suggests' field are needed to run this function.")
  }

  rmarkdown::render(
    input = file.path(tempdir(), "qc_idats.Rmd"),
    output_file = output_file,
    output_dir = output_directory,
    encoding = encoding,
    params = list(
      csv_file = csv_file,
      data_directory = data_directory,
      array = array,
      annotation = annotation,
      filter_snps = filter_snps,
      filter_non_cpg = filter_non_cpg,
      filter_xy = filter_xy,
      filter_multihit = filter_multihit,
      filter_beads = filter_beads,
      population = population,
      bead_cutoff = bead_cutoff,
      detection_pvalues = detection_pvalues,
      filter_callrate = filter_callrate,
      callrate_samples = callrate_samples,
      callrate_probes = callrate_probes,
      gender_threshold = gender_threshold,
      gender_colname = gender_colname,
      norm_background = norm_background,
      norm_dye = norm_dye,
      norm_quantile = norm_quantile,
      cell_tissue = cell_tissue,
      pca = pca,
      pca_vars = pca_vars,
      pca_threshold = pca_threshold,
      max_labels = max_labels,
      title = title,
      author_name = author_name,
      author_affiliation = author_affiliation,
      author_email = author_email,
      output_directory = output_directory
    ),
    ...
  )

  message(message_prefix, "Quality-Control ended.")

  invisible()
}
